src/base/path.cpp

```diff
@@ -0,0 +1,326 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "path.h"
+
+#include <QDataStream>
+#include <QDir>
+#include <QFileInfo>
+#include <QMimeDatabase>
+#include <QRegularExpression>
+#include <QVector>
+
+#if defined(Q_OS_WIN)
+const Qt::CaseSensitivity CASE_SENSITIVITY = Qt::CaseInsensitive;
+#else
+const Qt::CaseSensitivity CASE_SENSITIVITY = Qt::CaseSensitive;
+#endif
+
+const int PATHLIST_TYPEID = qRegisterMetaType<PathList>();
+
+Path::Path(const QString &pathStr)
+    : m_pathStr {QDir::cleanPath(pathStr)}
+{
+}
+
+Path::Path(const std::string &pathStr)
+    : Path(QString::fromStdString(pathStr))
+{
+}
+
+Path::Path(const char pathStr[])
+    : Path(QString::fromLatin1(pathStr))
+{
+}
+
+bool Path::isValid() const
+{
+    if (isEmpty())
+        return false;
+
+#if defined(Q_OS_WIN)
+    const QRegularExpression regex {QLatin1String("[:?\"*<>|]")};
+#elif defined(Q_OS_MACOS)
+    const QRegularExpression regex {QLatin1String("[\\0:]")};
+#else
+    const QRegularExpression regex {QLatin1String("[\\0]")};
+#endif
+    return !m_pathStr.contains(regex);
+}
+
+bool Path::isEmpty() const
+{
+    return m_pathStr.isEmpty();
+}
+
+bool Path::isAbsolute() const
+{
+    return QDir::isAbsolutePath(m_pathStr);
+}
+
+bool Path::isRelative() const
+{
+    return QDir::isRelativePath(m_pathStr);
+}
+
+bool Path::exists() const
+{
+    return !isEmpty() && QFileInfo::exists(m_pathStr);
+}
+
+Path Path::rootItem() const
+{
+    const int slashIndex = m_pathStr.indexOf(QLatin1Char('/'));
+    if (slashIndex < 0)
+        return *this;
+
+    if (slashIndex == 0) // *nix absolute path
+        return createUnchecked(QLatin1String("/"));
+
+    return createUnchecked(m_pathStr.left(slashIndex));
+}
+
+Path Path::parentPath() const
+{
+    const int slashIndex = m_pathStr.lastIndexOf(QLatin1Char('/'));
+    if (slashIndex == -1)
+        return {};
+
+    if (slashIndex == 0) // *nix absolute path
+        return (m_pathStr.size() == 1) ? Path() : createUnchecked(QLatin1String("/"));
+
+    return createUnchecked(m_pathStr.left(slashIndex));
+}
+
+Path Path::filename() const
```

Comment:
>I think it should return `QString` instead.

Filename component of path is still path, isn't it? We can easily do something like `Path path = someDirPath / otherPath.filename();`

> This can avoid (IMO) confusing code pattern of `filename().data()`

Ok. I'll try to check whether it add other confusing code instead.

Reply:
>Filename component of path is still path, isn't it? 

IMO file name is more (naturally) like a string. That string can become `Path` but then it gains additional meaning: it is a relative path to a filename.